### PR TITLE
BDD: Graql Steps

### PR DESCRIPTION
**Pull request by alexjpwalker**
## What is the goal of this PR?

We had Connection and Concept BDD implementations, but Graql steps were still not implemented. We added support for them.

## What are the changes implemented in this PR?

- Implement all Graql Steps (BDD)
- Fix bad IID format in `Thing` (no leading `0x`)

### CODE REVIEW

File: grakn/concept/proto/concept_proto_builder.py [hunk 1]
```diff
@@ -96,7 +96,7 @@ def value_type(value_type_: ValueType):
 
 
 def iid(iid_: str):
-    return bytes.fromhex(iid_)
+    return bytes.fromhex(iid_[2:])
```
Comment (alexjpwalker) [inline grakn/concept/proto/concept_proto_builder.py:99]: Previously, the IID did not include `0x`, which is required for Grakn and Graql to parse it correctly. We added it.

File: tests/behaviour/context.py [hunk 2]
```diff
@@ -17,21 +17,38 @@
...
+from grakn.concept.type.type import Type
 from grakn.rpc.session import Session
 from grakn.rpc.transaction import Transaction
 
 
-ThingSubtype: Thing = Union[Entity, Relation, BooleanAttribute, LongAttribute, DoubleAttribute, StringAttribute, DateTimeAttribute]
+AttributeSubtype: Attribute = Union[BooleanAttribute, LongAttribute, DoubleAttribute, StringAttribute, DateTimeAttribute]
+ThingSubtype: Thing = Union[Entity, Relation, AttributeSubtype]
+TypeSubtype: Type = Union[ThingType, EntityType, RelationType, RoleType, AttributeType, BooleanAttributeType, LongAttributeType, DoubleAttributeType, StringAttributeType, DateTimeAttributeType]
+ConceptSubtype: Concept = Union[ThingSubtype, TypeSubtype]
```
Comment (alexjpwalker) [inline tests/behaviour/context.py:50]: I looked for a more elegant way of expressing these, and tried `typing.Type`- but the trouble with that is that it includes the type itself that you specify, so it was thinking, for example, that `ConceptSubtype` could be a `Concept`, which is untrue.

File: grakn/concept/proto/concept_proto_builder.py (hunk 1 above)
Comment (vmax) [reply grakn/concept/proto/concept_proto_builder.py:99]: @alexjpwalker AFAIR, you're trying to remove '0x' part here? If so, the intention might be clearer if you do:

```suggestion
    return bytes.fromhex(iid_.lstrip('0x'))
```

File: tests/behaviour/graql/graql_steps.py [hunk 3]
```diff
@@ -16,108 +16,370 @@
...
+        else:
+            raise TypeError("A Concept was matched by label, but it is not a Type.")
+
+
+class AttributeMatcher(ConceptMatcher):
+
+    def __init__(self, type_and_value: str):
+        s = type_and_value.split(":")
+        assert_that(s, has_length(2), "[%s] is not a valid attribute identifier. It should have format \"type_label:value\"." % type_and_value)
+        self.type_label = s[0]
```
Comment (vmax) [inline tests/behaviour/graql/graql_steps.py:164]: @alexjpwalker this and next line could be shortened to:

```suggestion
        self.type_label, self.value = s
```

File: tests/behaviour/graql/graql_steps.py [hunk 4]
```diff
@@ -16,108 +16,370 @@
...
+        for key in keys:
+            if key.as_remote(context.tx()).get_type().get_label() == self.type_label:
+                return self.check(key)
+
+        return False
+
+
+def parse_concept_identifier(value: str):
+    concept_identifier_parts: List[str] = value.split(":", 1)
+    identifier_type = concept_identifier_parts[0]
```
Comment (vmax) [inline tests/behaviour/graql/graql_steps.py:213]: @alexjpwalker 

```suggestion
    identifier_type, identifier_body = concept_identifier_parts
```

File: tests/behaviour/graql/graql_steps.py [hunk 5]
```diff
@@ -16,108 +16,370 @@
...
+    elif identifier_type == "key":
+        return ThingKeyMatcher(type_and_value=identifier_body)
+    elif identifier_type == "value":
+        return AttributeValueMatcher(type_and_value=identifier_body)
+    else:
+        raise ValueError("Failed to parse concept identifier: " + value)
+
+
+def answer_concepts_match(context: Context, answer_identifier: List[Tuple[str, str]], answer: ConceptMap) -> bool:
+    for (var, concept_identifier) in answer_identifier:
```
Comment (vmax) [inline tests/behaviour/graql/graql_steps.py:226]: @alexjpwalker 

```suggestion
    for var, concept_identifier in answer_identifier:
```

File: tests/behaviour/graql/graql_steps.py [hunk 6]
```diff
@@ -16,108 +16,370 @@
...
+
+@step("uniquely identify answer concepts")
+def step_impl(context: Context):
+    answer_identifiers = parse_table(context.table)
+    assert_that(context.answers, has_length(len(answer_identifiers)),
+                "The number of answers [%d] should match the number of answer identifiers [%d]." % (len(context.answers), len(answer_identifiers)))
+
+    result_set = [(ai, []) for ai in answer_identifiers]
+    for answer in context.answers:
+        for (answer_identifier, matched_answers) in result_set:
```
Comment (vmax) [inline tests/behaviour/graql/graql_steps.py:242]: @alexjpwalker 

```suggestion
        for answer_identifier, matched_answers in result_set:
```

File: tests/behaviour/graql/graql_steps.py [hunk 7]
```diff
@@ -16,108 +16,370 @@
...
+            if answer_concepts_match(context, answer_identifier, answer):
+                matched_answers.append(answer)
+
+    for (answer_identifier, answers) in result_set:
```
Comment (vmax) [inline tests/behaviour/graql/graql_steps.py:246]: @alexjpwalker 

```suggestion
    for answer_identifier, answers in result_set:
```

File: tests/behaviour/graql/graql_steps.py [hunk 8]
```diff
@@ -16,108 +16,370 @@
...
+    assert_that(context.numeric_answer is not None, reason="The last query executed was not an aggregate query.")
+    assert_numeric_value(context.numeric_answer, expected_answer)
+
+
+@step("aggregate answer is not a number")
+def step_impl(context: Context):
+    assert_that(context.numeric_answer.is_nan())
+
+
+class AnswerIdentifierGroup(object):
```
Comment (vmax) [inline tests/behaviour/graql/graql_steps.py:292]: @alexjpwalker as we're now Python3-only, we can omit `(object)` inheritance from everywhere

**Review by vmax — APPROVED** [review]

**Pull request merged**